Add diff highlighting line transformer

janosh opened the pull request:
Closes #62.

This is a draft implementation of a line transformer that enables diff highlighting on code fences with the [`{diff}` option](https://github.com/andrewbranch/gatsby-remark-vscode/pull/88/files#diff-6887d09d163fd4e53b30ba2ab651343fR1).

- Lines starting with `+ `&thinsp;indicate additions and can be highlighted with a green transparent background.
- Lines starting with `- `&thinsp;indicate deletions and can be highlighted with a red transparent background.

This seems like a very common use case to me so I took the liberty of [including `createDiffLineTransformer` in `getDefaultLineTransformers`](https://github.com/andrewbranch/gatsby-remark-vscode/pull/88/files#diff-0797f8b54e6ed443744041853b3e38efR16). Of course, feel free to undo that if you think this is more niche.

Also, due to my complete lack of JSDoc knowledege, I'd like to appeal to [`CONTRIBUTING.md`](https://github.com/andrewbranch/gatsby-remark-vscode/blob/master/CONTRIBUTING.md) which states:
> If types are new to you, or you have trouble getting type checking to pass, that’s ok! I’ll help you in your PR if necessary.

Specifically, I'm having trouble with the [type declaration on the transformer returned by `createDiffLineTransformer`](https://github.com/andrewbranch/gatsby-remark-vscode/pull/88/files#diff-939de1f342ea398f64eba69a09dae4b0R6).

Finally (not sure if the infrastructure for this is implemented yet), [how do I access a code fence's`parsedOptions` in a line transformer](https://github.com/andrewbranch/gatsby-remark-vscode/pull/88/files#diff-939de1f342ea398f64eba69a09dae4b0R7)?

```js
const transformer = ({ line, parsedOptions: { diff } }) => { ... }
```

ain't working.

andrewbranch commented:
Cool! Looking forward to taking a look soon!

janosh commented:
I've time to work on this some more tomorrow in case you already have some feedback.

andrewbranch commented:
Hey @janosh, sorry for the delay. I started to respond last weekend, but as I was reviewing, I realized there’s a problem, and I’ve been trying to think of a solution. First of all, though, thanks so much for continuing to contribute to this plugin!

So, the problem is that leaving the `+` and `-` characters in the source will mess up the syntax analysis, because they’ll be interpreted as part of the source code (e.g. as binary or unary operators in JavaScript), which will incorrectly influence the classification of the tokens the precede.

The only solution is to strip off those characters in the line transformer, and add them back with CSS or a custom renderer. (This would also solve the lesser problem of the characters affecting apparent indentation relative to lines that don’t have a +/- prefix, as the +/- character could be rendered in a gutter instead of pushing the rest of the line out to the right.) But, it seems like it could be difficult to do that in a way that looks right automatically for every theme. The same goes for the green/red line background color. You’ve set it to transparent by default, requiring users to pick colors that work for their theme(s), which matches the current behavior of the line highlighting transformer—but, via #51, I really want to change that and try to do some smart color analysis so highlighting works automatically. It seems like a strategy like that would be more difficult for green and red highlights, but I do want to get away from using transparent as a default in the future, because it makes it seem like the default experience is broken. These are tough problems.

The direction I’m leaning is that if we can’t provide defaults that work well for most users out of the box, I’d rather the transformer be a separate package that users can install, add to the plugin options, and configure however they need to. (If you decide to move this code to your own package, I’ll still be happy to help get it working as I think it’s good for the ecosystem.)

What are your thoughts?

janosh commented:
@andrewbranch To me it sounds like you've already devised adequate solutions to all the problems you mention.

> The only solution is to strip off those characters in the line transformer, and add them back with CSS or a custom renderer.

Adding them into the gutter with CSS would be good, I think. That seems to be exactly what VS Code does as well.
<img width="1459" alt="Screen Shot 2020-03-28 at 07 22 52" src="https://user-images.githubusercontent.com/30958850/77816706-1f0c7600-70c5-11ea-90a7-7cece0ac01c7.png">

A custom renderer might be more hassle than is necessary.

> But, it seems like it could be difficult to do that in a way that looks right automatically for every theme.

I don't see why. Could you elaborate?

>  but I do want to get away from using transparent as a default in the future, because it makes it seem like the default experience is broken.

Yes, I was thinking the same thing while creating this PR. But I forgot about #51 and that you're already planning to change this.

I checked how red and green highlighting looks on different dark backgrounds while working on this. Looked great to me with every theme I checked. And in fact, if you open a diff view in VS Code and cycle through the pre-installed dark themes, you'll see that all except one (High Contrast) use similar shades of green/red transparent backgrounds for highlighting added/removed lines. And the plus/minus signs in the gutter of added/removed lines are exactly identical across all themes. So I think just setting `--grvsc-line-diff-(add|del)-background-color` to the values in [`test/integration/viewer/render.js`](https://github.com/andrewbranch/gatsby-remark-vscode/pull/88/files#diff-5e9fc378436b37f9aabfd0845cda82f8R54-R55) would do the job. But since people already specify a VS Code theme in the extension's config, shouldn't there be a way of adopting whichever values the theme specifies for `--grvsc-line-diff-(add|del)-background-color`?

One piece of functionality in VS Code that's missing entirely from this PR is that lines that are only modified are highlighted with a more transparent shade of green/red than lines that are completely new/deleted. Those get a stronger green/red background (see screenshot above). But figuring out the "newness" of a line would require significant code analysis and likely add a huge amount of complexity to this PR so I'm thinking we should not do that.

andrewbranch commented:
Sorry for the delay. A couple responses:

>> But, it seems like it could be difficult to do that in a way that looks right automatically for every theme.
>
>I don't see why. Could you elaborate?

Sorry, I think when I typed that, my brain went off in one direction and my fingers just typed “theme” instead. I think what I _meant_ to say was that it could be difficult to introduce an ad-hoc gutter that wouldn’t disrupt any custom styles the user has put on the code block. Like, it’s tempting just to absolutely position a `+` or `-` floating on top of the line to the far left without changing the overall structure of how lines are laid out, but whether you have room to do that, and how far from the border you place it, would depend on the horizontal padding, which is configurable via CSS variable. I can easily see that variable set very small on mobile viewports, which wouldn’t leave enough room for the +/- character. That strategy could also interfere with line numbers (#28), which I think is in higher demand for an included line transformer. In a world where there’s both a line number option and a diff option, they would clearly need to work together in order not to step on each others toes.

> I checked how red and green highlighting looks on different dark backgrounds while working on this. Looked great to me with every theme I checked. And in fact, if you open a diff view in VS Code and cycle through the pre-installed dark themes, you'll see that all except one (High Contrast) use similar shades of green/red transparent backgrounds for highlighting added/removed lines.

Ok, this is good to hear. That eliminates the color concern for me. If we can figure out how to deal with the gutter without introducing a ton of complexity, I’ll feel good about adding this.

janosh commented:
> If we can figure out how to deal with the gutter without introducing a ton of complexity, I’ll feel good about adding this.

That's great! I'll think about this tomorrow.

janosh commented:
How do you feel about allowing line transformers to perform some layout consistency checks? By that I mean the diff line transformer could check if the CSS variable for left padding is equal to or larger than a minimum value, say 16px. If not the case, it could either issue a warning to the user or just increase it to the minimum required value (in the hopes that this might be less annoying to users).

Taking a step back and reconsidering the need for diff icons, I don't think I've ever consciously noted the little `+`/`-` signs in a diff view before starting work on this PR. I feel like the red and green highlighting already provide sufficient visual guidance on what's been added and removed. Even color blind people probably just assume that what's highlighted in a different shade of gray below another line whose content is quite similar constitutes new code that replaces the old above. Not sure there's a strong need for `+`/`-` even there.

In general though, I don't see how you could safely insert any line prefixes, be they line numbers or diff icons (`+`/`-`) without checking for it in the line transformer itself. It alone knows how much horizontal padding it needs.

andrewbranch commented:
> How do you feel about allowing line transformers to perform some layout consistency checks? By that I mean the diff line transformer could check if the CSS variable for left padding is equal to or larger than a minimum value, say 16px. If not the case, it could either issue a warning to the user or just increase it to the minimum required value (in the hopes that this might be less annoying to users).

I don’t see how this is possible—you won’t have access to CSS (or DOM functions like `getComputedStyle`) at build time.

I think it might be worth it to make the gutter a first-class concept. I’m thinking that a line transformer can optionally return some text/html to insert into the gutter, and then if any line from a given line transformer returns non-empty gutter text, we’ll push a preceding span onto every line. By setting them to `display: table-cell`, we can ensure that the contributions from different line transformers are all aligned, and the overall gutter width will be consistent across every line, but as narrow as possible: https://codepen.io/atcb/pen/ZEbabYL. It still needs a little work, as this display mode breaks the existing method of setting horizontal padding.

I’ll plan to work on this tomorrow and/or over the course of next week in a separate PR. I think that should set us up well for this PR and for line numbers. Thanks for your patience on this!

janosh commented:
> I don’t see how this is possible—you won’t have access to CSS (or DOM functions like `getComputedStyle`) at build time.

Oh right, that's a good point. I keep forgetting about this. 🤦 

You could maybe make this work with Gatsby's Server Rendering APIs ([onPreRenderHTML](https://www.gatsbyjs.org/docs/ssr-apis/#onPreRenderHTML) and [onRenderBody](https://www.gatsbyjs.org/docs/ssr-apis/#onRenderBody)) but it would probably be a pain to make those available to line transformers. Even if you manage, post-processing the user's CSS sounds brittle and like a certain path to unhappiness. 😞

> I think it might be worth it to make the gutter a first-class concept.

👍

andrewbranch commented:
Update: did a big chunk of work on this today, but `display: table-cell` didn’t work out as well as I thought it would, so need to iterate a little more on the rendering and layout. Because this is going to change the element structure (and so could break people’s CSS), I’m planning on rolling all this into a v3.

janosh commented:
> Because this is going to change the element structure (and so could break people’s CSS), I’m planning on rolling all this into a v3.

Sounds good. Let me know if there's anything I can help test.

andrewbranch commented:
If you want to rebase onto the `v3` branch (and by that I mean both `git rebase` your branch and change the base branch for this PR from `master` to `v3`), you should be unblocked! In your line transformer, you’ll want to strip the leading `+` or `-` from the result text, and instead return that character in the new `gutterCells` property. You can use the line number transformer I added as an example: https://github.com/andrewbranch/gatsby-remark-vscode/blob/cb774a22e04386ee82d3d3209f026c61ea2dbfff/src/transformers/lineNumberTransformer.js#L31-L36

After you rebase, I can take a closer look at your initial questions about accessing the code fence `parsedOptions` in your line transformer, although the lineNumberTransformer above does that, so maybe that will help. Thanks for your patience as always; I’m excited to finally be able to get this rolling again!

janosh commented:
@andrewbranch Very cool! Thanks for all the effort you put into this plugin.

I rebased, changed the base branch to `v3` and used the new `gutterCells` prop in `diffLineTransformer`. So is `meta` the 'new' `parsedOptions`?

janosh commented:
@andrewbranch Still need to update the tests but maybe it's good to wait until the code is ready. As always, feel free to step in and make changes to this PR wherever you want.

andrewbranch commented on src/transformers/diffLineTransformer.js:
```diff
@@ -0,0 +1,47 @@
+  /** @type {LineTransformer<any>} */
+  const diffLineTransformer = ({ line, meta: { diff } }) => {
+    if (diff && line.text.startsWith('+ ')) {
+      line.text = line.text.replace(/^\+ /, '');
```
I _think_ it doesn’t matter at this point, but it would probably be best to avoid mutating the input parameters. I should type them as `readonly`, in fact.

janosh replied:
Changing to:

```js
const text = line.text.replace(/^\+ /, '');
return {
  line: highlightDiffLine({ ...line, text }, 'add'),
```

andrewbranch commented on src/transformers/diffLineTransformer.js:
```diff
@@ -0,0 +1,47 @@
+  // How do I get the parsed code fence options here?
+  /** @type {LineTransformer<any>} */
+  const diffLineTransformer = ({ line, meta: { diff } }) => {
+    if (diff && line.text.startsWith('+ ')) {
```
GitHub’s diff grammar doesn’t require the space... do you think we would get false positives if we didn’t require it? I can’t think of many times where an actual line of source code begins with a `+` (without any leading indentation), but surely some rare cases exist. I wonder if it would also be worth it to allow escapes with `\+`... probably not, but I think we should consider dropping the space.

janosh replied:
Should we still strip off a space after the `+/-` (if there is one)? Or just leave it as is?

```js
const text = line.text.replace(/^\+/, '');   // or
const text = line.text.replace(/^\+ ?/, '');
return {
  line: highlightDiffLine({ ...line, text }, 'add'),
```

andrewbranch replied:
If a space isn’t necessary to trigger the highlight, then a space should not be stripped.

andrewbranch replied:
Wait, on second thought, this seems fine. One space can get stripped. This is a lot like Markdown’s treatment of the optional space after block quote markers (`>`).

janosh replied:
> This is a lot like Markdown’s treatment of the optional space after block quote markers (`>`).

Exactly, that's what I was thinking. Not checking for a space in the `if(diff && line.text.startsWith('-'))` statement anymore but still stripping one (if available) now.

andrewbranch commented on src/transformers/diffLineTransformer.js:
```diff
@@ -0,0 +1,47 @@
+  diffLineTransformer.displayName = 'diff';
+  diffLineTransformer.schemaExtension = `
+    type GRVSCTokenizedLine {
+      isDiff: Boolean
```
This should match what’s in `data`, so looks like it should be

```suggestion
      isDiffAdd: Boolean
      isDiffDel: Boolean
```

or maybe an enum would be better:

```graphql
enum GRVSCDiff {
  ADD
  DEL
}

type GRVSCTokenizedLine {
  diff: GRVSCDiff
}
```

(and update the `data` property with the literal string `"ADD"` or `"DEL"`)

andrewbranch replied:
Hmm, maybe I can also try to generate TypeScript typings from these extensions so it can be validated properly...

janosh replied:
I added the enum.

andrewbranch commented on styles.css:
```diff
@@ -66,7 +66,16 @@
+}
+
+.grvsc-line-diff-add {
+  background-color: var(--grvsc-line-diff-add-background-color, transparent);
```
Since we’re putting this in v3, go ahead and pick default colors that you think will work well 🎉 

With the layout changes I made, though, I think you’ll need to use a pseudoelement like I changed line highlighting to use. Reason is that the `.grvsc-line` span is no longer guaranteed to be as wide as the container, unfortunately, so this will result in the highlight stopping at the end of the text instead of extending all the way to the right edge of the container.

I think if a line is both highlighted and diff-highlighted, the diff highlight should have priority. Do you think that makes sense? (I expect this is a severe edge case, but probably something that should be tested/handled nonetheless.)

andrewbranch replied:
Re pseudoelement: you can reuse the existing `.grvsc-has-line-highlighting` container class to generate the `::before` element, then set its background-color on `.grvsc-line-diff-add::before`.

janosh replied:
> Since we’re putting this in v3, go ahead and pick default colors that you think will work well 🎉

I went with `rgba(0, 255, 60, 0.4)` (additions) and `rgba(255, 0, 20, 0.4)` (deletions) for now. They are similar (but not identical) to the colors used on GitHub.

> I think if a line is both highlighted and diff-highlighted, the diff highlight should have priority. Do you think that makes sense?

Sounds sensible.

> Re pseudoelement: you can reuse the existing `.grvsc-has-line-highlighting` container class to generate the `::before` element, then set its background-color on `.grvsc-line-diff-add::before`.

I tried. Not sure I did this right. Let me know if something's off.

andrewbranch replied:
<img width="622" alt="image" src="https://user-images.githubusercontent.com/3277153/82272822-86f67480-9930-11ea-9f01-201ce9599812.png">

Looks right!

andrewbranch commented:
> As always, feel free to step in and make changes to this PR wherever you want.

I took the liberty of updating the tests, fixing the one type error (nice job on the rest of the types by the way), and updating the postfix space requirement locally. But it looks like you don’t have the “allow edits from maintainers” checkbox checked 😁

janosh commented:
> ... and updating the postfix space requirement locally.

Oops, should have read this first.

> But it looks like you don’t have the “allow edits from maintainers” checkbox checked 😁

I _do_ have it checked:

<img width="649" alt="Screen Shot 2020-05-19 at 07 47 00" src="https://user-images.githubusercontent.com/30958850/82289783-d0cd7200-99a5-11ea-8077-311f0d7d6ecc.png">

I tried unchecking and rechecking just now. Maybe that helps?

andrewbranch commented:
> I do have it checked:

Weird, I’m still getting permission denied. Maybe it's because the PR is from your master branch? I’ll PR my changes to your remote, then. I think it’s ready to merge after that.

andrewbranch commented:
Welp, looks like GitHub is having downtime, so I guess I’m not doing this right now 😐

andrewbranch commented:
Ok, after refreshing 100 times I created it: https://github.com/janosh/gatsby-remark-vscode/pull/1

janosh commented:
@andrewbranch My pleasure! Do you have a timeline for v3? Let me if I can help.

andrewbranch commented:
No strict timeline, depends on how long everything takes in #96, but I plan on spending some time on it this weekend. I think the most time-consuming task will be writing good documentation for the GraphQL stuff and line transformers. Finishing this weekend (even with the US holiday on Monday) is probably too optimistic. Probably by the end of next weekend is the best-case scenario.

You can always give `gatsby-remark-vscode@alpha` a try in the meantime, and let me know if anything listed in #96 sounds like something you want to work on, although most of it is just housekeeping and docs writing at this point.
